File: app/src/main/java/de/bornholdtlee/defaultproject/utils/Logger.java
package de.bornholdtlee.defaultproject.utils;

import android.util.Log;

import de.bornholdtlee.defaultproject.BuildConfig;

public final class Logger {

    private Logger() {
    }

    public static void error(String message) {
        if (isAllowedToLog()) {
            Log.e(getCallingClass(), message);
        }
    }

    public static void debug(String message) {
        if (isAllowedToLog()) {
            Log.d(getCallingClass(), message);
        }
    }

    public static void info(String message) {
        if (isAllowedToLog()) {
            Log.i(getCallingClass(), message);
        }
    }

    public static void verbose(String message) {
        if (isAllowedToLog()) {
            Log.v(getCallingClass(), message);
        }
    }

    public static void warn(String message) {
        if (isAllowedToLog()) {
            Log.w(getCallingClass(), message);
        }
    }

    public static void wtf(String message) {
        if (isAllowedToLog()) {
            Log.wtf(getCallingClass(), message);
        }
    }

    private static boolean isAllowedToLog() {
        return BuildConfig.DEBUG;
    }

    private static String getCallingClass() {
        StackTraceElement[] stackTraceElements = Thread.currentThread().getStackTrace();
        for (int i = 0; i < stackTraceElements.length; i++) {
            if (stackTraceElements[i].getClassName().equals(Logger.class.getCanonicalName())) {
                try {
                    String[] split = stackTraceElements[i + 2].getClassName().split("\\.");
                    String className = split[split.length - 1];

                    if (className.contains("$")) {
                        className = className.split("\\$")[0];
                    }

                    return "(" + className + ".java:" + stackTraceElements[i + 2].getLineNumber() + ")";
                } catch (Exception e) {
                    return Logger.class.getSimpleName();
                }
            }
        }
        return Logger.class.getSimpleName();
    }
}

File: app/src/main/java/de/bornholdtlee/baseproject/utils/Logger.kt
package de.bornholdtlee.baseproject.utils

import android.util.Log

import de.bornholdtlee.baseproject.BuildConfig

object Logger {

    private val isAllowedToLog: Boolean
        get() = BuildConfig.DEBUG

    private val callingClass: String
        get() {
            val stackTraceElements = Thread.currentThread().stackTrace
            for (i in stackTraceElements.indices) {
                if (stackTraceElements[i].className == Logger::class.java.canonicalName) {
                    return try {
                        val split = stackTraceElements[i + 2].className.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                        var className = split[split.size - 1]

                        if (className.contains("$")) {
                            className = className.split("\\$".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0]
                        }

                        "(" + className + ".java:" + stackTraceElements[i + 2].lineNumber + ")"
                    } catch (e: Exception) {
                        Logger::class.java.simpleName
                    }

                }
            }
            return Logger::class.java.simpleName
        }

    fun error(message: String) {
        if (isAllowedToLog) {
            Log.e(callingClass, message)
        }
    }

    fun debug(message: String) {
        if (isAllowedToLog) {
            Log.d(callingClass, message)
        }
    }

    fun info(message: String) {
        if (isAllowedToLog) {
            Log.i(callingClass, message)
        }
    }

    fun verbose(message: String) {
        if (isAllowedToLog) {
            Log.v(callingClass, message)
        }
    }

    fun warn(message: String) {
        if (isAllowedToLog) {
            Log.w(callingClass, message)
        }
    }

    fun wtf(message: String) {
        if (isAllowedToLog) {
            Log.wtf(callingClass, message)
        }
    }
}